:robot: Audit: Auto-update with latest upstream patches (#194)
sehlen-bsi/botan-docs: docs/audit_report/src/02_security_issues.rst

These side channels are present in Botan 3.0.0 to 3.2.0 and were fixed in
|botan_version|. The fix essentially hard-codes the division by `q` as a series
of constant-time bit shift and multiplication operations. See also :ref:`changes/fixes`.

.. _secinfo/oversized_ecc:

Oversized elliptic curve parameters cause DoS
---------------------------------------------

For custom elliptic curves, Botan didn't reject an oversized parameter ``p``.

> Instead it would attempt to verify the primality of the parameter. Espcially for
> malicously large parameters with thousands of bits this could waste a
> significant amount of computation time. Starting with Botan |botan_version|, the

Reviewer: ```suggestion
Instead, it would attempt to verify the primality of the parameter. Especially for
maliciously large parameters with thousands of bits, this could waste a
significant amount of computation time. Starting with Botan |botan_version|, the
```

maximum size of ``p`` is limited to 1024 bits.

A vulnerable application would have to invoke one of the following ``EC_Group``
constructors or factory methods with parameters from an untrusted source:

* ``EC_Group::EC_Group(std::string_view pem_or_oid)``
* ``EC_Group::EC_Group(const uint8_t& ber[], size_t ber_len)``
* ``EC_Group::EC_Group(const std::vector<uint8_t>& ber)``
* ``EC_Group::EC_Group_from_PEM(std::string_view pem)``.

> Please explicitly note that the all of the above constructors accept both an

Reviewer: ```suggestion
Please explicitly note that all of the above constructors accept both an
```

encoded OID (of a named curve) or encoded parameters of a custom curve. The
disambiguation is performed internally. This could allow an attacker to
masquerade a custom curve as an OID of a named curve and exploit the
vulnerability.

> Note that these constructors are called also internally and may receive

Reviewer: ```suggestion
Note that these constructors are also called internally and may receive
```

user-provided or potentially untrusted data. For instance, when loading a
maliciously crafted key in an ASN.1 container or even embedded into an X.509
certificate.

> For a vulnerable server requiring TLS client authentication this

Reviewer: ```suggestion
For a vulnerable server requiring TLS client authentication, this
```

might be a feasible denial of service amplifier without requiring the attacker

> to be in a privileged network location. An attacker on the network path, could
> mount a similar attack by impersonating the server and sending a malicious

Reviewer: ```suggestion
to be in a privileged network location. An attacker on the network path could
mount a similar attack by impersonating the server and sending a malicious
```

  parameters in its `parameters` field.
* ``X509::load_key(...)`` (all of its overloads), with a custom curve encoded
  into a key container format. This is used to load public keys embedded in
  X.509 certificates.

This issue was reported to the maintainer by Bing Shi.

.. _secinfo/binary_gcd:

Faulty gcd() implementation
---------------------------

As shown by the OSS-Fuzz project, Botan's previous implementation to find the
greatest common denominator of two arbitrarily sized integers generated an error
for certain inputs. The GCD algorithm was based on "safegcd" [SAFEGCD]_, but
with a slight adaption to avoid the requirement of constant-time arithmetics on
signed integers. This adaption violated the algorithm's termination assumptions,
causing the sporadic error.

Most notably, the gcd() function is used in the RSA key generation. As a result,
key generation might have sporadically failed with an exception. In the worst
case this could have caused a bias in the RSA key pairs generated by Botan.
However, we are not aware of any failures in a production environment.

As a permanent fix, Botan |botan_version| uses the binary GCD algorithm

> [BINARYGCD]_, which does not require constant-time arithmetics on signed intgers

Reviewer: ```suggestion
[BINARYGCD]_, which does not require constant-time arithmetics on signed integers
```

and provides a termination assumption that is much easier to reason about.

> Additionally, the gcd() now leaks less information via timing side channels as
> before.

Reviewer: This sounds a bit like we fixed some of the many side channels. Maybe we want to write something like:
```suggestion
Additionally, the new gcd() is more resistant against timing side channels than the previous version.
```
